odin.qmd

```diff
@@ -0,0 +1,299 @@
+
+:::
+
+::::
+
+. . .
+
+* `dt` is a special parameter
+* every variable has `initial` and `update` pair
+
+
+## ...compared with ODE models
+
+:::: {.columns}
+
+::: {.column style="font-size: 75%"}
+```{r}
+#| echo: false
+#| results: "asis"
+r_output(readLines("models/sir-basic.R"))
+```
+:::
+
+::: {.column style="font-size: 60%"}
+```{r}
+#| echo: false
+#| results: "asis"
+r_output(readLines("models/sir-basic-ode.R"))
+```
+
+:::
+
+::::
+
+
+## Compiling with `odin2`
+
+```{r}
+#| echo: false
+#| results: "asis"
+r_output(model_compile_code("sir", "models/sir-basic.R"))
+```
+
+```{r}
+#| include: false
+sir <- odin2::odin("models/sir-basic.R")
+```
+
+
+## Running a single simulation
+
+
+```{r}
+sys <- dust2::dust_system_create(sir, pars = list(), dt = 0.25)
+dust2::dust_system_set_state_initial(sys)
+t <- seq(0, 100)
+y <- dust2::dust_system_simulate(sys, t)
+y <- dust2::dust_unpack_state(sys, y)
+
+plot(t, y$I, type = "l", xlab = "Time", ylab = "Infected population")
+```
+
+
+## Running multiple simulations
+
+```{r}
+sys <- dust2::dust_system_create(sir, pars = list(), n_particles = 50,
+                                 dt = 0.25)
+dust2::dust_system_set_state_initial(sys)
+t <- seq(0, 100)
+y <- dust2::dust_system_simulate(sys, t)
+y <- dust2::dust_unpack_state(sys, y)
+matplot(t, t(y$I), type = "l", lty = 1, col = "#00000044",
+        xlab = "Time", ylab = "Infected population")
+```
+
+
+## Calculating incidence with `zero_every`
+
+```{r}
+#| echo: false
+#| results: "asis"
+r_output(model_compile_code("sir", "models/sir-incidence.R"))
+```
+
+```{r}
+#| include: false
+sir <- odin2::odin("models/sir-incidence.R")
+```
+
+
+## Incidence accumulates then resets
+
+```{r}
+sys <- dust2::dust_system_create(sir, pars = list(), dt = 1 / 128)
+dust2::dust_system_set_state_initial(sys)
+t <- seq(0, 20, by = 1 / 128)
+y <- dust2::dust_system_simulate(sys, t)
+y <- dust2::dust_unpack_state(sys, y)
+
+plot(t[t %% 1 == 0], y$incidence[t %% 1 == 0], type = "l",
+     ylab = "Infection incidence", xlab = "Time")
+lines(t, y$incidence, col = "red")
+```
+
+
+## Using `interpolate`
+
+The `interpolate` function in odin can be used for time-varying parameters, with specification of
+
+- the times of changepoints
+- the values at those changepoints
+- the type of interpolation: linear, constant or spline
+
+## Using `interpolate`
+
+```{r}
+#| echo: false
+#| results: "asis"
+r_output(model_compile_code("sir", "models/sir-interpolate-constant.R"))
+```
+
+```{r}
+#| include: false
+sir <- odin2::odin("models/sir-interpolate-constant.R")
+```
+
+
+## Using `interpolate`
+
+```{r}
+pars <- list(beta_time = c(0, 30), beta_value = c(0.2, 1))
+sys <- dust2::dust_system_create(sir, pars = pars, dt = 0.25)
+dust2::dust_system_set_state_initial(sys)
+t <- seq(0, 100)
+y <- dust2::dust_system_simulate(sys, t)
+y <- dust2::dust_unpack_state(sys, y)
+
+plot(t, y$I, type = "l", xlab = "Time", ylab = "Infected population")
+```
+
+
+## Using arrays
```

Comment:
This section definitely could do with expansion, but we start dropping off the edge of what can be fit onto a slide very quickly.  Let's discuss strategy this week